src/ChangeStream.php

```diff
@@ -58,36 +48,53 @@ public function current()
 
     public function getId()
     {
-        return $this->csIt->getId();
+        return $this->csIt->getInnerIterator()->getId();
+    }
+
+    public function key()
+    {
+
     }
 
     public function next()
     {
         try {
             $this->csIt->next();
-            $this->resumeToken = $this->csIt->extract_resume_token($this->csIt->current());
-        } catch (ResumableException $e) {
+            $this->resumeToken = $this->extractResumeToken($this->csIt->current());
+        } catch (RuntimeException $e) {
             $this->resume();
         }
     }
 
     public function resume()
     {
-        $this->options['resumeAfter'] = $this->resumeToken;
-        array_shift($this->pipeline);
-
-        $server = $this->manager->selectServer($this->options['readPreference']);
-
-        $command = new ChangeStreamCommand($this->databaseName, $this->collectionName, $this->pipeline, $this->options, $this->manager);
-        $server = $this->manager->selectServer($this->options['readPreference']);
-        $cursor = $command->resume($server, $this->pipeline);
-
-        $this->csIt = new ChangeStreamIterator($cursor);
+        $newChangeStream = call_user_func($this->resumeCallable, $this->resumeToken);
+        $this->csIt = $newChangeStream->csIt;
         $this->csIt->rewind();
     }
 
     public function rewind()
     {
         $this->csIt->rewind();
     }
+
+    public function valid()
+    {
+
+    }
+
+    private function extractResumeToken($document)
+    {
+        if ($document === null) {
+            throw new ResumeTokenException("Cannot extract a resumeToken from an empty document");
+        }
+        if ($document instanceof Serializable) {
+            return $this->extractResumeToken($document->bsonSerialize());
+        }
+        $lastResumeToken = is_array($document) ? $document['_id'] : $document->_id;
```

Comment:
Is there a case where we'd want to leave `$this->resumeToken` untouched if extraction fails? Would we expect users to catch this ResumeTokenException and try to carry on with iteration? I suppose they'd just hit another extraction error in that case since the cursor data wouldn't have changed.